lib/modules/datasource/generic.ts

```diff
@@ -0,0 +1,102 @@
+import is from '@sindresorhus/is';
+import jsonata from 'jsonata';
+import { z } from 'zod';
+import { logger } from '../../logger';
+import { readLocalFile } from '../../util/fs';
+import { parseUrl } from '../../util/url';
+import { Datasource } from './datasource';
+import type { ReleaseResult } from './types';
+
+export const ReleaseResultZod = z.object({
+  releases: z.array(
+    z.object({
+      version: z.string(),
+      isDeprecated: z.boolean().optional(),
+      releaseTimestamp: z.string().optional(),
+      sourceUrl: z.string().optional(),
+      sourceDirectory: z.string().optional(),
+      changelogUrl: z.string().optional(),
+    })
+  ),
+  sourceUrl: z.string().optional(),
+  sourceDirectory: z.string().optional(),
+  changelogUrl: z.string().optional(),
+  homepage: z.string().optional(),
+});
+
+export abstract class GenericDatasource extends Datasource {
+  override caching = true;
+  override customRegistrySupport = true;
+
+  protected async queryRegistry(
+    registryUrl: string | null | undefined
+  ): Promise<string | null> {
+    if (is.nullOrUndefined(registryUrl)) {
+      return null;
+    }
+
+    const url = parseUrl(registryUrl);
+    if (is.nullOrUndefined(url)) {
+      logger.debug(`Failed to parse url ${registryUrl}`);
+      return null;
+    }
+
+    let response: string | null = null;
+    switch (url.protocol) {
+      case 'http:':
+      case 'https:':
+        response = (await this.http.get(registryUrl)).body;
+        break;
+      case 'file:':
+        response = await readLocalFile(
+          registryUrl.replace('file://', ''),
+          'utf8'
+        );
+        break;
+      default:
+        logger.debug(`Scheme ${url.protocol} is not supported`);
+        return null;
+    }
+
+    if (!is.nonEmptyString(response)) {
+      return null;
+    }
+
+    return response;
+  }
+
+  protected async parseData(
+    packageName: string,
+    data: unknown
+  ): Promise<ReleaseResult | null> {
+    const expression = jsonata(packageName);
+    // wildcard means same object
+    const evaluated =
+      packageName === '*' ? data : await expression.evaluate(data);
+
+    const parsed = ReleaseResultZod.safeParse(evaluated);
+    if (!parsed.success) {
+      logger.debug({ err: parsed.error }, 'Response has failed validation');
+      return null;
+    }
+
+    const cleanedResponse = parsed.data;
+    // manually copy to prevent leaking data into other systems
```

Comment:
Why not clone?

Reply:
Does ZOD prevent additional attributes?

Reply:
What do you mean? I just meant using `structuredClone` here instead of manually copying fields.